### ISSUE TITLE
Query join methods are not really useful

### ISSUE THREAD

**Issue by thejahweh** [yiisoft/yii2#10203]
You see in #10174 that the join methods are not really what someone should expect if he wants to make a join.

Normally joins are used like this:
```sql
SELECT *
FROM TableA
    INNER JOIN TableB on (TableB.aID = TableA.id)
    INNER JOIN TableC on (TableC.bID = TableB.id)
```
I would now expect that it is something like this in Yii2:
```php
$query = new Query();
$query->select('*')
    ->from('TableA')
    ->innerJoin('TableB', ['aID' => ['TableA', 'id']])
    ->innerJoin('TableC', ['bID' => ['TableB', 'id']]);
```
But this generates this query:
```sql
SELECT *
FROM [TableA]
    INNER JOIN [TableB] ON [aID] IN ('TableA', 'rubID')
    INNER JOIN [TableC] ON [bID] IN ('TableA', 'rubID')
```
Ok then comes the `joinWith` method, which is great but only if the join is on the first table. I would expect that this works:
```php
$query = new Query();
$query->select('*')
    ->from('TableA')
    ->joinWith('b') // Works
    ->joinWith('c'); // Doesn't, because TableA does not have a relation with TableC
```
But it doesn't.

So yea it is possible to accomplish it with Yii2, but it is more a ugly :pensive: than a beautiful Yii2-like way :heart:
```php
$query = new Query();
$query->select('*')
    ->from('TableA')
    ->innerJoin('TableB', ['TableB.aID' => new Expression('TableA.id')])
    ->innerJoin('TableC', ['TableC.bID' => new Expression('TableB.id')]);
```
This doesn't look that ugly, but could you imagine it with longer not that readable table names? Or if the table names are coming from the classes?

**Comment by quantum13** [comment]
It just incorrect usage of innerJoin and joinWith.
innerJoin (http://www.yiiframework.com/doc-2.0/yii-db-query.html#innerJoin()-detail):

> The join condition that should appear in the ON part. Please refer to where() on how to specify this parameter.


joinWith (http://www.yiiframework.com/doc-2.0/yii-db-activequery.html#joinWith()-detail):
>Sub-relations can also be specified (see with()).

i.e. it must be written as:
```php
A::find()->joinWith('b.c')->all();
```

**Comment by cebe** [comment]
@thejahweh what about this?

```php
->innerJoin('TableC', 'TableC.bID = TableB.id');
```

works and is readable.

**Comment by thejahweh** [comment]
@quantum13 Thanks, this works and looks beautiful :heart: but it is a bit hard to find in the docs.

@cebe Yes thats how I do it since some weeks. But this can get also ugly with the DB-specific escaping:
```php
->innerJoin(['usefulNameB' => $tableB::tableName()], '[usefulNameB].[id] = [usefulNameA].[bID]')
```

**Comment by starrychloe** [comment]
Yeah, I agree, not very useful. I would have expected that you could just write 

    Account::find()->join('campaign')->join('ad_group')

And it would use the existing relationship defined in the models. Instead, you have to keep repeating the ON condition over & over again throughout the code, when it is already defined in the model. 

    Account::find()->innerJoin('campaign', 'account_id = account.id')->innerJoin('ad_group', 'campaign_id = campaign.id')

Even if you use `joinWith` then you have to have to specify the `eagerLoad` and `INNER JOIN` over and over again, because the default is `LEFT JOIN`. 

Rails has `.joins` where you only have to give the name of the relation, and you can express nested joins with multiple tables. 

http://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-joins

http://guides.rubyonrails.org/active_record_querying.html#using-array-hash-of-named-associations
